pages/src/TodoList.js

```diff
@@ -51,28 +55,75 @@ const withState = provideState({
       ({
         ...state,
         newTodoLabel: value
+      }),
+    onDisplayChange: (_, value) => state =>
+      ({
+        ...state,
+        display: value
+      }),
+    onTodoCompletionChange: (effects, target) => state => {
+      const { id } = target.dataset
+      effects.updateTodo({
+        ...state.todos.get(id),
+        completed: target.checked
+      })
+      effects.sendToServer({
+        ...state.todos.get(id),
+        completed: target.checked
       })
+    }
   }
 })
 
+const getVisibleTodos = createSelector(
+  state => state.display,
+  state => state.todos,
+  (display, todos) => {
+    if (display === 'all') {
+      return todos.valueSeq()
+    }
+
+    return todos.valueSeq().filter(
+      display === 'active'
+        ? t => !t.completed
+        : t => t.completed
+    )
+  }
+)
+
 export const TodoList = ({ state, effects }) => (
 
   <div>
-    <form onSubmit={persistEvent(effects.onCreateTodo)}>
+    <form onSubmit={effects.onCreateTodo}>
       <p>
         <input
-          onChange={persistEventInputValue(effects.setNewTodoLabelChange)}
+          onChange={extractEventValue(effects.setNewTodoLabelChange)}
           placeholder='Type and press <Enter> to create an item'
           size='40'
           value={state.newTodoLabel}
         />
       </p>
     </form>
+    <select onChange={extractEventValue(effects.onDisplayChange)} value={state.display}>
+      <option value='all'>All</option>
+      <option value='active'>Active</option>
+      <option value='completed'>Completed</option>
+    </select>
     <ul>
-      {state.todos.valueSeq().map(({ id, label }) =>
+      {getVisibleTodos(state, state).map(({ completed, id, label }) =>
```

Comment:
J'ai un bug ici, je ne savais pas quoi mettre dans le deuxième paramètre du selector, je dois comprendre mieux Reselect

Reply:
The arguments of selector are passed to its input selectors.